### DISCUSSION TITLE
Support multiple slots of data in `ol/DataTile`

### DISCUSSION THREAD

**Discussion by kikuchan** [openlayers/openlayers#16088] [Ideas]
Hi,

I want to composite a tile image (served by OSM, for example) with a floating-point data tile, so that I can access both at the same time via the `band` operator of WebGL style expressions.
This requires multiple textures to be available within a shader at the same time. However, it seems that `ol/DataTile` and `ol/webgl/TileTexture` currently don't support uploading multiple data types at once.

To address this, I'm considering modifying the private `data_` attribute in `ol/DataTile` to be an array, which would require some corresponding changes.
Specifically, changing the signature of `options.loader` in `ol/DataTile` might affect compatibility, but I believe this can be managed for now.

What do you think about this change?

**Comment by kikuchan** [comment]
Eventually, I composed a draft PR #16090 using this approach. It seems to be working fine so far, as I expected.

**Comment by mike-000** [comment]
Are multiple data types really necessary?  A custom loader could convert sources of different types to a single type, e.g. https://stackblitz.com/edit/js-pryjyu?file=package.json,index.html,index.js

**Comment by kikuchan** [comment]
Hi @mike-000, thank you for your comment!

...Yeah, it could be done. Anything could be done with `DataTile`; it's an amazing and powerful abstraction, though. I like it.

However, it impacts performance. It requires re-ordering the data of tiles to pack into a single buffer while each buffer is already in a suitable format for a texture.
Moreover, properly manipulating the data of multiple images within a single buffer can be somewhat cumbersome.

IMHO, a loader of `ol/source/DataTile` (DataTile **source**) should also be able to return multiple slots for data, but I think it's too late to change that now.
